Objects/setobject.c

```diff
@@ -905,10 +906,24 @@ set_update_internal(PySetObject *so, PyObject *other)
         return 0;
     }
 
+    /* If there's a length or length hint, do one big resize at the start,
+     * rather than incrementally resizing as we insert new keys.
+     * Expect that there will be no (or few) overlapping keys.
+     * Use 0 as a default value to avoid changing the current mask.
+     */
+    length = PyObject_LengthHint(other, 0);
+    if (length < 0)
+        return -1;
+
     it = PyObject_GetIter(other);
     if (it == NULL)
         return -1;
 
+    if (length > so->mask) {
+        if (set_table_resize(so, (so->used + length)*2) != 0)
+            return -1;
+    }
+
```

Comment:
How about use 
```c
 if ((so->fill + length)*5 >= so->mask*3) {
            if (set_table_resize(so, (so->used + length)*2) != 0)
                return -1;
        }
```
copied from dict case?

Reply:
At least, I think these two cases should be consistent.

Reply:
OK, I'll do that.